boot/bootstate20.go

```diff
@@ -356,31 +356,103 @@ func (envbks *envRefExtractedKernelBootloaderKernelState) setNextKernel(sn snap.
 }
 
 //
-// kernel snap methods
+// bootStateUpdate for 20 methods
 //
 
-// bootState20Kernel implements the bootState and bootStateUpdate interfaces for
-// kernel snaps on UC20. It is used for setNext() and markSuccessful() - though
-// note that for markSuccessful() a different bootStateUpdate implementation is
-// returned, see bootState20MarkSuccessful
-type bootState20Kernel struct {
-	bks bootloaderKernelState20
+type bootCommitTask func() error
+
+// bootState20Kernel implements the bootStateUpdate interface for both kernel
+// and base snaps on UC20.
+type bootStateUpdate20 struct {
+	// tasks to run before the modeenv has been written
+	preModeenvTasks []bootCommitTask
+
+	// the modeenv that was read from the state
+	modeenv *Modeenv
+
+	// the modeenv that will be written out in commit
+	writeModeenv *Modeenv
+
+	// tasks to run after the modeenv has been written
+	postModeenvTasks []bootCommitTask
+}
+
+func newBootStateUpdate20(m *Modeenv) *bootStateUpdate20 {
+	u20 := &bootStateUpdate20{
+		writeModeenv: &Modeenv{},
+	}
+	// TODO: is this too silly?
+	u20.modeenv = m
+	dest := make([]Modeenv, 1)
+	copy(dest, []Modeenv{*m})
+	u20.writeModeenv = &dest[0]
+
+	return u20
+}
+
+// commit will write out boot state persistently to disk.
+func (u20 *bootStateUpdate20) commit() error {
+	// The actual actions taken here will depend on what things were called
+	// before commit(), either setNextBoot for a single type of kernel snap, or
+	// markSuccessful for kernel and/or base snaps.
+	// It is expected that the caller code is carefully analyzed to avoid
+	// critical points where a hard system reset during that critical point
+	// would brick a device or otherwise severely fail an update.
+	// There are three things that callers can do before calling commit(),
+	// 1. modify writeModeenv to specify new values for things that will be
+	//    written to disk in the modeenv.
+	// 2. Add tasks to run before writing the modeenv.
+	// 3. Add tasks to run after writing the modeenv.
+
+	// first handle any pre-modeenv writing tasks
+	for _, t := range u20.preModeenvTasks {
+		if err := t(); err != nil {
+			return err
+		}
+	}
+
+	// next write the modeenv
+	if !u20.modeenv.deepEquals(u20.writeModeenv) {
+
+		origin, err := u20.modeenv.Origin()
+		// should be impossible since we should have read the modeenv
+		if err != nil {
+			return fmt.Errorf("internal error: %v", err)
+		}
+		if err := u20.writeModeenv.WriteTo(origin); err != nil {
```

Comment:
isn't this block just Modeevn.Write?  a bit unclear to me why we need adding Origin

Reply:
yes, with the Duplicate method also copying the private originRootDir over, we are now able to just do `u20.writeModeenv.Write()` as it shares the same originRootDir as the modeenv we got originally.